README.md
# Simplex

[![Build Status](https://travis-ci.org/adamkittelson/simplex.svg)](https://travis-ci.org/adamkittelson/simplex)
[![Coverage Status](https://img.shields.io/coveralls/adamkittelson/simplex.svg)](https://coveralls.io/r/adamkittelson/simplex)

An Elixir library for interacting with the [Amazon SimpleDB](http://aws.amazon.com/simpledb/) API.

Requires Elixir ~> 1.0.0

## Installation

Install the [Hex.pm](http://hex.pm) package

1. Add simplex to your `mix.exs` dependencies:

    ```elixir
    def deps do
      [
        {:simplex, "0.3.0"}
      ]
    end
    ```

2. Add `:simplex` to your application dependencies:

    ```elixir
    def application do
      [applications: [:simplex]]
    end
    ```

### Configuration

`{:ok, simplex} = Simplex.new`

#### AWS Keys

To communicate with the SimpleDB API you'll need to provide your AWS Access and Secret keys.

There are two ways to provide your keys to the Simplex library:

1. Set them from within your application
    ```elixir
    {:ok, simplex} = Simplex.new("your-access-key", "your-secret-access-key")
    ```

    or

    ```elixir
    {:ok, simplex} = Simplex.new
    Simplex.aws_access_key(simplex, "your-access-key")
    Simplex.aws_secret_access_key(simplex, "your-secret-access-key")
    ```

2. Set them as the environment variables `AWS_ACCESS_KEY` and `AWS_SECRET_ACCESS_KEY`
    ```
    AWS_ACCESS_KEY=your-access-key AWS_SECRET_ACCESS_KEY=your-secret-access-key iex -S mix

    iex(1)> {:ok, simplex} = Simplex.new
    {:ok, #PID<0.164.0>}
    iex(2)> Simplex.aws_access_key(simplex)
    "your-access-key"
    iex(3)> Simplex.aws_secret_access_key(simplex)
    "your-secret-access-key"

    ```

3. If not provided by the above two methods Simplex will attempt to retrieve keys from [instance metadata](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html) if it's running in EC2 and you launched your instance with an IAM role with permission to access SimpleDB.

#### SimpleDB URL

By default Simplex will send all requests to the us-east-1 SimpleDB url: `https://sdb.amazonaws.com`. If you want to use a [different region](http://docs.aws.amazon.com/general/latest/gr/rande.html#sdb_region) you can change the url by:

1. Setting it from within your application
    ```elixir
    {:ok, simplex} = Simplex.new
    Simplex.simpledb_url(simplex, "https://sdb.us-west-1.amazonaws.com")
    ```

2. Set it as the environment variable `SIMPLEDB_URL`
    ```
    SIMPLEDB_URL=https://sdb.us-west-1.amazonaws.com iex -S mix

    iex(1)> {:ok, simplex} = Simplex.new
    {:ok, #PID<0.164.0>}
    iex(2)> Simplex.simpledb_url(simplex)
    "https://sdb.us-west-1.amazonaws.com"
    ```

#### SimpleDB Version

By default Simplex will use the 2009-04-15 version of the SimpleDB API. If you wish to use a different version you can change it by:

1. Setting it from within your application
    ```elixir
    {:ok, simplex} = Simplex.new
    Simplex.simpledb_version(simplex, "2008-02-12")
    ```

2. Set it as the environment variable `SIMPLEDB_VERSION`
    ```
    SIMPLEDB_VERSION=2008-02-12 iex -S mix

    iex(1)> {:ok, simplex} = Simplex.new
    {:ok, #PID<0.164.0>}
    iex(2)> Simplex.simpledb_version(simplex)
    "2008-02-12"
    ```


## Responses

Simplex will respond to SimpleDB requests with a 3 element tuple, either `{:ok, result, response}` or `{:error, messages, response}`

#### %Simplex.Response{}

A Simplex response (third element of the tuple above) has the following fields:

1. body: A Map containing the parsed body of the response
2. status_code: The HTTP status code of the response
3. headers: The response headers
4. raw_body: The raw string of the response body. (XML)

#### Pattern Matching

You can pattern match to determine how to handle the response:

  ```elixir
  case Simplex.Domains.create(simplex, "new_domain") do
    {:ok, result, response} ->
      # some happy path stuff here
    {:error, messages, response} ->
      # handle problems
  end
  ```

## Domains

[Create](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_CreateDomain.html) a new domain.

  ````elixir
  Simplex.Domains.create(simplex, "new_domain")
  ````

[List](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_ListDomains.html) domains.

  ````elixir
  Simplex.Domains.list(simplex)

  Simplex.Domains.list(simplex, %{"MaxNumberOfDomains" => "10", "NextToken" => "token-from-previous-list-response"})
  ````

[Delete](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_DeleteDomain.html) a domain.

  ````elixir
  Simplex.Domains.delete(simplex, "domain_to_delete")
  ````

## Attributes

[Get](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_GetAttributes.html) attributes of an item.

  ````elixir
  Simplex.Attributes.get(simplex, "your_domain", "your_item_name")

  Simplex.Attributes.get(simplex, "your_domain", "your_item_name", %{"AttributeName" => "some_attribute", "ConsistentRead" => "true"})
  ````

[Put](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_PutAttributes.html) attributes on an item (or create a new item).

  ````elixir
  # Attribute values can be strings, lists of strings, or a
  # two-element tuple of :replace and a string or list of strings
  # The replace tuple indicates that the value should replace the existing
  # value for that attribute, rather than be added to its values

  Simplex.Attributes.put(simplex,
                         "your_domain",
                         "your_item_name",
                         %{"some_key"        => "some_value",
                           "another_key"     => ["a", "list", "of", "values"],
                           "yet_another_key" => {:replace, "a value to replace the existing value(s) of yet_another_key"},
                           "one_last_key     => {:replace, ["values", "to", "replace", "one_last_key's", "previous", "value(s)"]}})

  # put "some_value" in the "some_key" attribute only if
  # "other_key" has the "other_value" value
  Simplex.Attributes.put(simplex,
                         "your_domain",
                         "your_item_name",
                         %{"some_key" => "some_value"},
                         %{"Name" => "other_key", "Value" => "other_value"})
  ````

[Delete](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_DeleteAttributes.html) attributes on an item (or delete an item).

  ````elixir
  # Delete the "some_value" value from the "some_key" attribute
  Simplex.Attributes.delete(simplex,
                            "your_domain",
                            "your_item_name",
                            %{"some_key" => "some_value"})

  # Delete "your_item_name" if it doesn't have the "some_key" attribute
  Simplex.Attributes.delete(simplex,
                            "your_domain",
                            "your_item_name",
                            %{},
                            %{"Name" => "some_key", "Exists" => "false"})
  ````

## Select

[Select](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_Select.html) attributes from items matching an expression.

  ````elixir
  Simplex.Select.select(simplex, "select * from your_domain where some_key = 'some_value'")
  ````

----

#### TODO:

1. Implement [BatchDeleteAttributes](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_BatchDeleteAttributes.html), [BatchPutAttributes](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_BatchPutAttributes.html), and [DomainMetadata](http://docs.aws.amazon.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_DomainMetadata.html)
2. Docs
3. Implement Select.select_all / Domains.list_all (automatically follow next_token to load more results if present)


CHANGELOG.md
--------------------------------------------------------------------------------
                        ^ ADD NEW CHANGES ABOVE ^
--------------------------------------------------------------------------------

CHANGELOG
=========

---- 0.0.2 / 2014-09-11 / configuration ----------------------------------------
* configurable aws creds and simpledb url

---- 0.0.3 / 2014-09-12 / docs-and-tests ---------------------------------------
* set up travis/coveralls

---- 0.0.4 / 2014-09-13 / readme -----------------------------------------------
* update readme

---- 0.1.0 / 2014-09-13 / ------------------------------------------------------
* initial release to hex.pm

---- 0.1.2 / 2014-09-17 / metadata-credentials ---------------------------------
* retrieve access keys from IAM roles when running in EC2

---- 0.1.3 / 2014-09-17 / refactor-config --------------------------------------
* allow multiple configs instead of one global config

---- 0.2.3 / 2014-09-18 / token ------------------------------------------------
* fix IAM credentials by including token

---- 0.2.5 / 2014-09-23 / configurable-version ---------------------------------
* allow configuration of simpledb version

---- 0.2.6 / 2014-10-09 / retry-on-error ---------------------------------------
* retry with exponential backoff for 500 errors

---- 0.2.7 / 2014-10-15 / update-httpoison -------------------------------------
* update httpoison

---- 0.2.8 / 2014-12-08 / httpotion --------------------------------------------
* switch to httpotion / ibrowse

---- 0.2.10 / 2015-04-01 / retry-httpotion-errors ------------------------------
* retry errors, upgrade httpotion, remove exvcr test dependency

---- 0.3.1 / 2015-07-10 / fix-key-refresh-date-comparison ----------------------
* fix date comparison that determines when aws creds need refreshing

---- 0.3.3 / 2015-12-16 / dep-updates ------------------------------------------
* update dependencies


lib/simplex.ex
defmodule Simplex do
  use GenServer
  use Timex

  def new do
    start_link(%{})
  end

  def new(access_key, secret_access_key) do
    start_link(%{:aws_access_key => access_key, :aws_secret_access_key => secret_access_key})
  end

  def start_link(config, options \\ []) do
    config = config
             |> Map.put_new(:aws_access_key, System.get_env("AWS_ACCESS_KEY"))
             |> Map.put_new(:aws_secret_access_key, System.get_env("AWS_SECRET_ACCESS_KEY"))
             |> Map.put_new(:simpledb_url, System.get_env("SIMPLEDB_URL") || "https://sdb.amazonaws.com")
             |> Map.put_new(:simpledb_version, System.get_env("SIMPLEDB_VERSION") || "2009-04-15")

    GenServer.start_link(__MODULE__, config, options)
  end

  def aws_access_key(simplex) do
    configuration(simplex)[:aws_access_key]
  end

  def aws_access_key(simplex, access_key) do
    GenServer.call(simplex, {:set_aws_access_key, access_key})
  end

  def aws_secret_access_key(simplex) do
    configuration(simplex)[:aws_secret_access_key]
  end

  def aws_secret_access_key(simplex, secret_access_key) do
    GenServer.call(simplex, {:set_aws_secret_access_key, secret_access_key})
  end

  def configuration(simplex) do
    GenServer.call(simplex, :get_configuration)
  end

  def simpledb_url(simplex) do
    configuration(simplex)[:simpledb_url]
  end

  def simpledb_url(simplex, url) do
    GenServer.call(simplex, {:set_simpledb_url, url})
  end

  def simpledb_version(simplex) do
    configuration(simplex)[:simpledb_version]
  end

  def simpledb_version(simplex, version) do
    GenServer.call(simplex, {:set_simpledb_version, version})
  end

  defp needs_refresh?(config) do
    expiring?(config) or missing_keys?(config)
  end

  defp missing_keys?(config) do
    !config[:aws_access_key] or !config[:aws_secret_access_key]
  end

  # keys expired or expiring within the next 60 seconds
  def expiring?(%{:expires_at => nil}), do: false
  def expiring?(%{:expires_at => expires_at}) do
    expires_at = DateFormat.parse!(expires_at, "{ISOz}")

    # Timex.Date.Compare/2:
    # -1 — the first date comes before the second one
    #  0 — both arguments represent the same date when coalesced to the same timezone.
    #  1 — the first date comes after the second one
    1 ==
      Date.now
      |> Date.shift(secs: 60)
      |> Date.compare(expires_at)
  end
  def expiring?(_config), do: false

  defp load_credentials_from_metadata do
   try do
      %HTTPotion.Response{:body => role_name} = HTTPotion.get("http://169.254.169.254/latest/meta-data/iam/security-credentials/", [timeout: 500])
      %HTTPotion.Response{:body => body} = HTTPotion.get("http://169.254.169.254/latest/meta-data/iam/security-credentials/#{role_name}", [timeout: 500])
      Poison.decode!(body)
    rescue
      _ ->
        %{}
    end
  end

  defp refresh(config) do
    update = load_credentials_from_metadata

    config
    |> Map.put(:aws_access_key,        update["AccessKeyId"]     || config[:aws_access_key])
    |> Map.put(:aws_secret_access_key, update["SecretAccessKey"] || config[:aws_secret_access_key])
    |> Map.put(:expires_at,            update["Expiration"]      || config[:expires_at])
    |> Map.put(:token,                 update["Token"]           || config[:token])
  end


  ###################
  # Server Callbacks

  def init(config) do
    {:ok, config}
  end

  def handle_call(:get_configuration, _from, config) do
    if needs_refresh?(config) do
      config = refresh(config)
      {:reply, config, config}
    else
      {:reply, config, config}
    end
  end

  def handle_call({:set_aws_access_key, access_key}, _from, config) do
    config = config
             |> Map.put(:aws_access_key, access_key)
             |> Map.delete(:expires_at)
             |> Map.delete(:token)
    {:reply, config[:aws_access_key], config}
  end

  def handle_call({:set_aws_secret_access_key, secret_access_key}, _from, config) do
    config = config
             |> Map.put(:aws_secret_access_key, secret_access_key)
             |> Map.delete(:expires_at)
             |> Map.delete(:token)
    {:reply, config[:aws_secret_access_key], config}
  end

  def handle_call({:set_simpledb_url, url}, _from, config) do
    config = Map.put(config, :simpledb_url, url)
    {:reply, config[:simpledb_url], config}
  end

  def handle_call({:set_simpledb_version, version}, _from, config) do
    config = Map.put(config, :simpledb_version, version)
    {:reply, config[:simpledb_version], config}
  end

  def handle_info(_msg, config) do
    {:noreply, config}
  end

end
